### PR TITLE
Create node.js.yml

### PR DESCRIPTION
**Pull request by ericrkuo**
Adding Continuous Integration 

Added new Node CI file https://docs.github.com/en/free-pro-team@latest/actions/guides/building-and-testing-nodejs
https://github.com/ericrkuo/Flow/blob/32fd557fc24ad898ae93c8bcb9a6071aa16938ce/.github/workflows/node.js.yml#L4

Added new secrets to hold our API keys for repository environment https://github.com/ericrkuo/Flow/settings/secrets/actions

This involved refactoring several classes so that our tests would pass
- Removed `RefreshCredential` from refreshing credential within **each** method in Spotify or Emotion
  - instead, we ensure that the caller ensures the credentials are refreshed before we call functions inside Spotify or Emotion
- Added `SpotifyApiWrapper` to catch rate limiting errors
  - if there are rate-limiting errors, wait 5 seconds before trying again
https://github.com/ericrkuo/Flow/blob/32fd557fc24ad898ae93c8bcb9a6071aa16938ce/service/SpotifyApiWrapper.js#L5-L23
- New folder structure
image

Note there aren't that many file changes. It's just a bunch of refactoring and renaming

### CODE REVIEW

**Review by ld357 — COMMENTED** [review]
Looks good! Thanks for all the refactoring! :) Just wanted to confirm that for the RefreshCredential, we're doing that check inside Main before every call to Emotion or Spotify?

**Comment by ericrkuo** [comment]
@ld357 do you mind re-reviewing this commit https://github.com/ericrkuo/Flow/pull/25/commits/ca8f6f718879e6be8337dad2670aaab04054c60d

this link may help https://stackoverflow.com/questions/63113333/is-there-a-way-to-call-a-function-before-calling-the-controllers-in-nodejs

File: main/src/routes/middleware.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        })
+        .catch((err) => {
+            throw err;
+        })
+}
+
+/**
+ * Checks whether the instance Main is valid
+ * */
+function isMainValid(main) {
```
Comment (ld357) [inline main/src/routes/middleware.js:51]: Since we're also checking the spotifyApi, refreshCredentialService, and the tokens, maybe a better name?

**Review by ld357 — APPROVED** [review]
Super clean! :) Just a small renaming thing but otherwise looks good to me!

**Pull request merged**